### PR TITLE
Custom entitlements: add README and other improvements

### PR DESCRIPTION
**Pull request by aboedo**
Added the README file and did more cleanup on other tings.

### CODE REVIEW

File: examples/CustomEntitlementComputationSample/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+always be calling configure only when the appUserID is already known. 
+- Update the applicationId in defaultConfig of the app-level build.gradle to match your RevenueCat app configuration.
+- Have at least one Offering with at least one Package configured for Android, since this is the one that the purchase button will use. 
+
+Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times CustomerInfoListener fired, as well as 
+the values for each one. 
+
+Happy testing!
+
+![sample screenshot](./Sample%20screenshot.png)
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:32]: just realized that this screenshot has my name, I'll update on Monday

File: examples/CustomEntitlementComputationSample/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+try {
+    val (transaction, customerInfo) =
+        Purchases.sharedInstance.awaitPurchase(purchaseParams)
+    // refresh entitlements with your backend
+} catch (error: PurchasesTransactionException) {
+    if (error.userCancelled) {
+        _uiState.update { it.copy(displayErrorMessage = "User cancelled") }
+    } else {
+        _uiState.update { it.copy(displayErrorMessage = error.message) }
+    }
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:99]: we should add the exact code to catch: 
- receipt already in use
- payment pending
- purchase cancelled
- product already purchased

File: examples/CustomEntitlementComputationSample/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+## Using Custom Entitlements mode
+
+### Installation: 
+
+This package is available on Maven and can be included via Gradle.
+
+To use this mode, ensure that you install the customEntitlementsComputation build flavor by specifying the dependency as:
+```gradle
+revenuecat = { module = "com.revenuecat.purchases:purchases@customEntitlementComputation", version.ref = "purchases" }
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:42]: could use some android-expert eyes on this

File: purchases/src/customEntitlementComputation/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt [hunk 4]
```diff
@@ -13,7 +13,6 @@ open class PurchasesConfiguration(builder: Builder) {
     internal val store: Store
     internal val diagnosticsEnabled: Boolean
     val dangerousSettings: DangerousSettings
-    val verificationMode: EntitlementVerificationMode
```
Comment (aboedo) [inline purchases/src/customEntitlementComputation/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt:16]: this was still available in this mode.

File: examples/CustomEntitlementComputationSample/app/src/main/java/com/revenuecat/sample/main/MainViewModel.kt [hunk 5]
```diff
@@ -15,8 +15,6 @@ import com.revenuecat.purchases.PurchasesTransactionException
 import com.revenuecat.purchases.awaitOfferings
 import com.revenuecat.purchases.awaitPurchase
 import com.revenuecat.purchases.interfaces.UpdatedCustomerInfoListener
-import com.revenuecat.purchases.purchaseWith
-import kotlinx.coroutines.coroutineScope
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/app/src/main/java/com/revenuecat/sample/main/MainViewModel.kt:19]: unused imports

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml [hunk 6]
```diff
@@ -2,7 +2,7 @@
 agp = "8.0.2"
 androidxNavigation = "2.5.3"
 kotlin = "1.7.20"
-purchases = "6.5.0"
+purchases = "6.8.0"
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:5]: not sure why this was 6.5.0

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml [hunk 7]
```diff
@@ -33,5 +33,4 @@ navigation-compose = "androidx.navigation:navigation-compose:2.5.3"
 compose-ui-tooling = { module = "androidx.compose.ui:ui-tooling" }
 compose-ui-test-manifest = { module = "androidx.compose.ui:ui-test-manifest" }
 
-revenuecat = { module = "com.revenuecat.purchases:purchases", version.ref = "purchases" }
-revenuecat-amazon = { module = "com.revenuecat.purchases:purchases-store-amazon", version.ref = "purchases" }
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:37]: why did we have amazon here? 
also, don't we have to specify the flavor?

File: purchases/src/customEntitlementComputation/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt (hunk 4 above)
Comment (tonidero) [reply purchases/src/customEntitlementComputation/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt:16]: Ah right... In iOS, we don't configure the SDK by passing a `PurchasesConfiguration`... Honestly, that feels less flexible (in case we ever do want to add optional features in the future) but it does disallow enabling these features... IMO, I think it's better to keep it this way for future-proofing, but lmk if you prefer the other way

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml (hunk 6 above)
Comment (tonidero) [reply examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:5]: I'm guessing since it is using the local version (in settings.gradle), it doesn't really matter?

File: examples/CustomEntitlementComputationSample/README.md (hunk 2 above)
Comment (tonidero) [reply examples/CustomEntitlementComputationSample/README.md:99]: Hmm they could check the `PurchasesTransactionException.error.errorCode` and check for the following error codes:
- `PurchasesErrorCode.ProductAlreadyPurchasedError`
- `PurchasesErrorCode.PurchaseCancelledError`: though it's easier to just use the `userCancelled` property.
- `PurchasesErrorCode.ReceiptAlreadyInUseError`: I might be missing something but this one shouldn't happen I think?  As in, we are posting purchases:
  -  Performed through the app (which should only be for new purchases and so should never have this error)
  -  On app foreground, if we haven't synced the purchase before. In this case, it would fail silently to sync, so a user in this mode would never see this happening.
- `PurchasesErrorCode.PaymentPendingError`: need to test to confirm this error happens...

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml (hunk 7 above)
Comment (tonidero) [reply examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:37]: Amazon now only has the `defaults` flavor, not the `customEntitlementComputation` flavor. But yeah, I don't think it's needed in this sample app.

File: examples/CustomEntitlementComputationSample/README.md (hunk 3 above)
Comment (tonidero) [reply examples/CustomEntitlementComputationSample/README.md:42]: This will have a different name once it's published. Also, this nomenclature is only if they use gradle catalogs... I would kinda vote to use simple gradle `implementation "com.revenuecat.purchases..."`

File: examples/CustomEntitlementComputationSample/README.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+```
+for Gradle version catalogs, or
+```gradle
+implementation 'com.revenuecat.purchases:purchases@customEntitlementsComputation:6.8.0'
```
Comment (vegaro) [inline examples/CustomEntitlementComputationSample/README.md:46]: `@customEntitlementsComputation` would be valid if we were publishing as a flavor, but we are publishing the flavor as a different artifact

```
implementation 'com.revenuecat.purchases:purchases-custom-entitlement-computation:6.8.0'
```

I would also remove the Gradle catalogs version and simply leave the standard nomenclature

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml (hunk 6 above)
Comment (vegaro) [reply examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:5]: It is outdated on MagicWeatherCompose, which is what I used as a base project for the sample. I am not sure why it's outdated there, but I guess it's an issue in the releasing scripts

Comment (tonidero) [reply examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:5]: Both magic weather sample apps don't update their version automatically... We definitely should fix that, even if it's just pointing to the latest snapshot...  I created a task [here](https://linear.app/revenuecat/issue/SDK-3242/auto-update-versions-in-magicweather-sample-apps-android).

File: examples/CustomEntitlementComputationSample/settings.gradle [hunk 9]
```diff
@@ -15,9 +15,9 @@ dependencyResolutionManagement {
 rootProject.name = "CustomEntitlementComputationSample"
 include ':app'
 
+// Uncomment to use local version of the SDK
 includeBuild("../../") {
     dependencySubstitution {
-        substitute module("com.revenuecat.purchases:purchases") using project(":purchases")
-        substitute module("com.revenuecat.purchases:purchases-store-amazon") using project(":feature:amazon")
+        substitute module("com.revenuecat.purchases:purchases@customEntitlementComputation") using project(":purchases")
```
Comment (tonidero) [inline examples/CustomEntitlementComputationSample/settings.gradle:21]: Should this be commented out now?

File: examples/CustomEntitlementComputationSample/Sample screenshot.png
Comment (tonidero) [inline examples/CustomEntitlementComputationSample/Sample screenshot.png:1]: I think Andy also wanted to take another screenshot without his name :P

**Review by tonidero — APPROVED** [review]
Looks great, left a few comments, but nothing blocking

File: examples/CustomEntitlementComputationSample/gradle/libs.versions.toml [hunk 10]
```diff
@@ -33,5 +33,4 @@ navigation-compose = "androidx.navigation:navigation-compose:2.5.3"
...
+revenuecat = { module = "com.revenuecat.purchases:purchases@customEntitlementComputation", version.ref = "purchases" }
```
Comment (tonidero) [inline examples/CustomEntitlementComputationSample/gradle/libs.versions.toml:36]: We should update this to use the correct package once it's deployed... I guess that should be in a different PR

Comment (vegaro) [reply examples/CustomEntitlementComputationSample/gradle/libs.versions.toml]: I forgot to update it here. I just did in [883202b](https://github.com/RevenueCat/purchases-android/pull/1167/commits/883202b3eee18e4375876d625aad4505fb0f4023)

File: examples/CustomEntitlementComputationSample/README.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
+# Custom Entitlements Computation Example App
+
+This app is useful for testing RevenueCat under Custom Entitlements Computation mode and understanding how it works.
+
+## What is Custom Entitlements Computation mode? 
+
+This is a special behavior mode for RevenueCat SDK, is intended for apps that will do their own entitlement computation separate from RevenueCat.
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:7]:
```suggestion
This is a special behavior mode for RevenueCat SDK. It's intended for apps that will do their own entitlement computation separate from RevenueCat. 
```

File: examples/CustomEntitlementComputationSample/README.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+This app is useful for testing RevenueCat under Custom Entitlements Computation mode and understanding how it works.
+
+## What is Custom Entitlements Computation mode? 
+
+This is a special behavior mode for RevenueCat SDK, is intended for apps that will do their own entitlement computation separate from RevenueCat. 
+
+Apps using this mode rely on webhooks to signal their backends to refresh entitlements with RevenueCat.
+
+In this mode, RevenueCat will not generate anonymous user IDs, it will not refresh customerInfo cache automatically only when a purchase goes through
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:11]:
```suggestion
In this mode, RevenueCat will not generate anonymous user IDs, or refresh customerInfo cache automatically. CustomerInfo is only refreshed when a purchase goes through 
```

File: examples/CustomEntitlementComputationSample/README.md [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+## What is Custom Entitlements Computation mode? 
+
+This is a special behavior mode for RevenueCat SDK, is intended for apps that will do their own entitlement computation separate from RevenueCat. 
+
+Apps using this mode rely on webhooks to signal their backends to refresh entitlements with RevenueCat.
+
+In this mode, RevenueCat will not generate anonymous user IDs, it will not refresh customerInfo cache automatically only when a purchase goes through 
+and it will disallow all methods other than those for configuration, switching users, getting offerings and making purchases.
+
+When in this mode, the app should use switchUser() to switch to a different App User ID if needed.
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:14]:
```suggestion
When in this mode, the app should use `switchUser()` to switch to a different App User ID if needed. 
```

File: examples/CustomEntitlementComputationSample/README.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+To use the app, you should do the following: 
+- Configure your app in the [RevenueCat dashboard](https://app.revenuecat.com/). No special configuration is needed, but you should contact RevenueCat support
+before enabling this mode to ensure that it's the right one for your app. It's highly recommended to set Transfer Behavior to "Keep with original App User ID" in the RevenueCat Dashboard. 
+- Update the API key in Constants.kt. You can update the default `appUserID` there too, although apps in this mode should 
+always be calling configure only when the appUserID is already known. 
+- Update the applicationId in defaultConfig of the app-level build.gradle to match your RevenueCat app configuration.
+- Have at least one Offering with at least one Package configured for Android, since this is the one that the purchase button will use. 
+
+Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times CustomerInfoListener fired, as well as
```
Comment (aboedo) [inline examples/CustomEntitlementComputationSample/README.md:27]:
```suggestion
Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times `CustomerInfoListener` fired, as well as 
```

File: examples/CustomEntitlementComputationSample/README.md (hunk 1 above)
Comment (aboedo) [reply examples/CustomEntitlementComputationSample/README.md:32]: we can use this one
Screenshot 2023-07-24 at 2 34 30 PM

File: examples/CustomEntitlementComputationSample/README.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                "The receipt is already in use by another subscriber. " +
+                        "Log in with the previous account or contact support " +
+                        "to get your purchases transferred to regain access"
+            PurchasesErrorCode.PaymentPendingError ->
+                "The purchase is pending and may be completed at a later time. " +
+                        "This can happen when awaiting parental approval or going " +
+                        "through extra authentication flows for credit cards " +
+                        "in some countries"
+            PurchasesErrorCode.ProductAlreadyPurchasedError ->
+                "This product is already active for the user."
```
[the remark this answers is not in the document]
Comment (aboedo) [reply examples/CustomEntitlementComputationSample/README.md]: sounds good.

File: examples/CustomEntitlementComputationSample/README.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        _uiState.update { it.copy(displayErrorMessage = errorMessage) }
+
+    }
+}
+```
+
+### Observing changes to purchases:
+
+To ensure that your app reacts to changes to subscriptions in real time, you can use `customerInfoStream`. This stream will only fire when new `customerInfo` is registered
+in RevenueCat, like when a subscription is renewed. If there are no changes from the last value, it will not fire. This means it's not guaranteed to fire on every app open.
```
[the remark this answers is not in the document]
Comment (aboedo) [reply examples/CustomEntitlementComputationSample/README.md:126]: yeah, good call, we should update

[the remark this answers is not in the document]
Comment (vegaro) [reply examples/CustomEntitlementComputationSample/README.md:126]: also, `customerInfoStream` is not a thing in Android. I will update

[the remark this answers is not in the document]
Comment (vegaro) [reply examples/CustomEntitlementComputationSample/README.md]: check it out and let me know what you think

File: examples/CustomEntitlementComputationSample/README.md (hunk 1 above)
Comment (vegaro) [reply examples/CustomEntitlementComputationSample/README.md]: Took another one with more padding in the `CustomerInfo listener values` section and included it

**Review by tonidero — APPROVED** [review]
